### PR TITLE
Diesel integration

### PR DESCRIPTION
**Pull request by Sytten**
## PR Info

See discussion: https://github.com/SeaQL/sea-query/discussions/168

**BLOCKED BY https://github.com/diesel-rs/diesel/pull/3691 and https://github.com/diesel-rs/diesel/pull/3694**

## New Features

- [x] Diesel binding

### CODE REVIEW

**Comment by Sytten** [comment]
@weiznich I would greatly appreciate if you could let me know if you think this is a good approach 🙏

**Comment by Sytten** [comment]
@tyt2y3 @billy1624 This is ready for a first review, we still have to wait for the next diesel release before we can merge though

File: sea-query-diesel/src/query.rs
Comment (weiznich) [inline sea-query-diesel/src/query.rs:25]: It might be useful to only call this method depending on whether sea-query believes that a prepared statement is meaningful to cache or not. It will significantly improve performance for queries that are executed more than once.

Diesel generally tries to cache all queries that are somewhat static, which means all queries that do not contain a user input dependent component (beside specific bind parameters).

**Review by weiznich — COMMENTED** [review]
From the diesel side the implementation looks OK. The one thing I personally would try to improve is enabling prepared statement caching at least for some queries at it really helps with performance.

File: sea-query-diesel/src/query.rs
Comment (weiznich) [reply sea-query-diesel/src/query.rs:25]: It might be already worth to have a simple heuristics there: There are no bound paramaters -> cache the query, as it does not change for different calls.

For more targeted caching you would need to access the query ast somehow to check if there is for example no `IN` expression with a variable number of binds in there or no `INSERT` statement with a variable number of rows.

Comment (weiznich) [reply sea-query-diesel/src/query.rs:25]: Also there is quite a lot of detailed documentation of diesels query caching here: https://docs.diesel.rs/2.1.x/diesel/connection/statement_cache/index.html#a-primer-on-prepared-statement-caching-in-diesel

Comment (tyt2y3) [reply sea-query-diesel/src/query.rs:25]: Thanks for the info. I will keep this in mind, and later may be we can have a API that walks the AST and returns false if:

1) any custom expression used
2) any `in` and `not_in` clause
3) insert many

**Comment by Sytten** [comment]
@tyt2y3 builds are failing because the bigdecimal versions dont match without a Cargo.lock, considering what I discussed in #664 are you OK with me adding a lock file to the examples and the crate?

**Comment by tyt2y3** [comment]
> considering what I discussed in #664 are you OK with me adding a lock file to the examples and the crate?

Sure, I am okay with it

**Comment by Sytten** [comment]
@tyt2y3 Ready for review

File: .github/workflows/rust.yml
Comment (weiznich) [inline .github/workflows/rust.yml:144]: This shouldn't be required anymore as diesel 2.1.1 added support for bigdecimal 0.4: https://crates.io/crates/diesel/2.1.1/dependencies

Comment (Sytten) [reply .github/workflows/rust.yml:144]: Ha the issue is on sea-query side, they only support 0.3 and the resolver takes 0.4 for diesel so it causes a build failure because the types are incompatible. Its a PITA that we still don't have a way to pin a sub dependency without a lockfile.

**Review by tyt2y3 — APPROVED** [review]
Thank you

**Comment by tyt2y3** [comment]
Released in https://crates.io/crates/sea-query-diesel/0.1.0

**Comment by Sytten** [comment]
Thanks much appreciated!

**Pull request merged**
